include/util/indexed_data.hpp

```diff
@@ -0,0 +1,366 @@
+        }
+    }
+};
+
+template <int N, typename T = std::string> struct FixedGroupBlock
+{
+    static constexpr std::uint32_t BLOCK_SIZE = N;
+
+    typedef T ResultType;
+    typedef typename T::value_type ValueType;
+
+    static_assert(sizeof(ValueType) == 1, "data basic type must char");
+
+    struct BlockReference
+    {
+        std::uint32_t offset;
+    };
+
+    FixedGroupBlock() {}
+
+    /// Write a block reference {offset}, where offset is a global block offset
+    /// Returns the fixed block prefix length.
+    template <typename Offset, typename OffsetIterator>
+    Offset
+    WriteBlockReference(std::ostream &out, Offset data_offset, OffsetIterator, OffsetIterator) const
+    {
+        BOOST_ASSERT(data_offset <= std::numeric_limits<decltype(BlockReference::offset)>::max());
+
+        BlockReference refernce{static_cast<decltype(BlockReference::offset)>(data_offset)};
+        out.write((const char *)&refernce, sizeof(refernce));
+
+        return BLOCK_SIZE;
+    }
+
+    /// Write a fixed length block prefix.
+    template <typename OffsetIterator>
+    void WriteBlockPrefix(std::ostream &out, OffsetIterator first, OffsetIterator last) const
+    {
+        std::uint32_t index = 0;
+        std::array<ValueType, BLOCK_SIZE> block_prefix;
+        for (OffsetIterator curr = first, next = std::next(first); curr != last; ++curr, ++next)
+        {
+            const std::uint32_t data_length = *next - *curr;
+            if (data_length >= 0x100)
+                throw util::exception(boost::format("too large data length %1%") % data_length);
+
+            block_prefix[index++] = static_cast<ValueType>(data_length);
+        }
+        out.write((const char *)block_prefix.data(), block_prefix.size());
+    }
+
+    /// Advances the range to an item stored in the referenced block.
+    /// Input [first..last) is a range of the complete block data with prefix.
+    /// Output [first..last) is a range of the referenced data at local_index.
+    template <typename DataIterator>
+    void ReadRefrencedBlock(const BlockReference &,
+                            std::uint32_t local_index,
+                            DataIterator &first,
+                            DataIterator &last) const
+    {
+        DataIterator fixed_lengths = first; // iterator to the fixed lengths part
+        std::advance(first, BLOCK_SIZE);    // advance first to the block data part
+        for (std::uint32_t i = 0; i < local_index; ++i)
+        {
+            first += static_cast<unsigned char>(*fixed_lengths++);
+        }
+
+        if (local_index < BLOCK_SIZE)
+        {
+            last = first + static_cast<unsigned char>(*fixed_lengths);
+        }
+    }
+};
+
+template <typename GroupBlock> struct IndexedData
+{
+    static constexpr std::uint32_t BLOCK_SIZE = GroupBlock::BLOCK_SIZE;
+
+    typedef std::uint32_t BlocksNumberType;
+    typedef std::uint64_t DataSizeType;
+
+    typedef typename GroupBlock::BlockReference BlockReference;
```

Comment:
nit pick: I always find the new using syntax easier to read: `using BlockReference = typename GroupBlock::BlockReference;`